department-of-veterans-affairs/caseflow
# Hun jer bah/appeals 9367 - Wrapped Error Handling in Split Appeal Controller

Resolves [APPEALS-9367](https://vajira.max.gov/browse/APPEALS-9367)

### Description
Create a wrapper within the split appeal controller that will revert any changes if there is an error during the split appeal process. The wrapper should also override the HearingDayFull error and allow an appeal to be split even if the hearing day is full.

### Acceptance Criteria
- [ ] When clicking "Split Appeal" on the Review and Confirm page of Split Appeal Workflow form:

- [ ] Issues in the form's new issue list are assigned to the new appeal created in [APPEALS-4147](https://vajira.max.gov/browse/APPEALS-4147)

- [ ] Issues in the form's not in the new issue list are maintained on the original appeal

- [ ] Issue assignments are wrapped in the same transaction as work done in [APPEALS-4156](https://vajira.max.gov/browse/APPEALS-4156) so if any step of splitting an appeal fails all changes are rolled back

### Testing Plan
1. Test that the wrapper works, you will need to force an error to occur during the split appeal process. The easiest way to do this is to force the `HearingDayFull` error to occur. Go to the **hearing.rb** model and disable line 113 within `check_available_slots`. Underneath, add `fail HearingDayFull if true` to force the HearingDayFull error. 
![image](https://user-images.githubusercontent.com/99915461/197260931-6f68584d-1b87-4c84-9e28-fd685d2e52ff.png)


2. Navigate to Caseflow and log in as a SSC user. 
![image](https://user-images.githubusercontent.com/99915461/197259357-d75204a5-9d92-4313-b8dd-ed27fde3def0.png)

3. Navigate to **Queue** and click on the **Search cases** link in the upper right corner. Look up docket number **221016-2** and select the case from the list. 
![image](https://user-images.githubusercontent.com/99915461/197259990-a281a09f-5e5e-44d6-99e9-cd06f319021a.png)

4. Click on the **correct issues** link to add issues to the appeal. 
![image](https://user-images.githubusercontent.com/99915461/197260080-64df5de8-07b8-4ec1-bd41-7a3a700581d9.png)

5. Click **Add Issue** and fill out the form that appears. Do this twice so the appeal has 2 issues on it. Click **save**.
![image](https://user-images.githubusercontent.com/99915461/197260284-6eb0b3cc-c7f9-44f4-9701-68bb1b09b335.png)
![image](https://user-images.githubusercontent.com/99915461/197260333-32094788-8b09-45a5-af89-a003af1a9124.png)
![image](https://user-images.githubusercontent.com/99915461/197260363-2c25990b-af4a-4ee9-adf2-c0c94bcefc64.png)

6. Click **correct issues** again to return to the edit screen 
![image](https://user-images.githubusercontent.com/99915461/197260452-cba79267-aedb-43b2-8d36-dc011e0d5ba3.png)

7. Click **Split Appeal** button.
![image](https://user-images.githubusercontent.com/99915461/197260503-1b7d0e7c-528d-4bca-9a9d-a0c6580a3919.png)

8. Fill out the split reason form and select a reason to continue. 
![image](https://user-images.githubusercontent.com/99915461/197260575-196e92f2-d78c-4a08-9bfd-2bcb269d93b8.png)

9. Click **Split appeal** to attempt to split the appeal
![image](https://user-images.githubusercontent.com/99915461/197260666-6979a5e0-319b-437d-96fb-11d9eb55cd92.png)

10. The banner should show that the split did not work. Behind the scenes, the appeal split failed at the hearing duplication because the hearing day was full. 
![image](https://user-images.githubusercontent.com/99915461/197261396-98969405-6b57-46b2-b0cb-b8fd1c85084d.png)

11. Check that the appeal wasn't created by copying the docket number and clicking the **search cases** link. Put the docket number into the search. There should only be the original appeal and no duplicate appeal. 

12. To confirm that the workflow is working properly, uncomment the code on line 113 of the `hearing` model from line 1 and remove any changes. 
![image](https://user-images.githubusercontent.com/99915461/197261844-ef6e4a08-b37e-4c30-9823-f75eb827b423.png)

13. Repeat steps 4-9 to split the appeal. This time, after clicking the **split appeal** button on the review page, you should be greeted with a green success banner!
![image](https://user-images.githubusercontent.com/99915461/197262261-867ba4db-5696-4a3f-a801-79b6abf82e45.png)

14. Copy the docket number and return to the **search cases** link. The split appeal should appear when you search the docket number.
![image](https://user-images.githubusercontent.com/99915461/197262000-80750b88-b82a-4bfc-b9dc-8648ad818139.png)
